cpp/src/fil/fil.cu

```diff
@@ -71,43 +71,63 @@ __global__ void transform_k(float* preds, size_t n, output_t output,
 }
 
 struct forest {
-  void init_max_shm() {
+  void enumerate_n_items() {
+    for (ssp_.n_items = algo_ == algo_t::BATCH_TREE_REORG ? 4 : 1;
+         ssp_.n_items != 0 &&
+         ssp_.max_shm < (ssp_.shm_sz = get_smem_footprint(ssp_, leaf_algo_));
```

Comment:
This condition is unnecessarily complex, and contains an assignment. Consider only checking for `n_items > 0` in the loop header, and perform the rest in the loop body.